# KEN 3140 Semantic Web: Lab 4
### Writing and executing basic SPARQL queries on RDF graphs

##### Learning objectives:
1. How to compose and execute basic SPARQL SELECT queries to answer a specific question
2. Identify and describe the function of the basic components of a SPARQL SELECT query
3. Given a question, identify what SPARQL features would be needed in order to answer it, and how to compose these features into a syntactically valid SPARQL query

#### SPARQL overview
* SPARQL (pronounced sparkle) stands for: **SPARQL Protocol And RDF Query Language**
* SPARQL 1.0 W3C-Recommendation since January 15th 2008
* SPARQL 1.1 W3C-Recommendation since March 21st 2013 Query language to query instances in RDF documents

**Overview of all SPARQL specifications: https://www.w3.org/TR/sparql11-overview/**

**Specification for the SPARQL query language: https://www.w3.org/TR/sparql11-query/**

> w3.org materials are standards and recommendations purveyed by the World Wide Web Consortium (W3C) - i.e., the body that defines Internet information standards

# SPARQL endpoint 🔗

* (Relational) databases are not built to be publicly available
* Triplestores, on the other hand, can be queried through a publicly available SPARQL endpoint URL
* Some Triplestore management systems enable user management, but SPARQL endpoints are natively built to be open and give the same access to all their users

We will explore the **DBpedia SPARQL endpoint** available at this link: 

>**https://dbpedia.org/sparql**

[DBpedia](https://wiki.dbpedia.org/) is a project to represent (parts of) Wikipedia as RDF, it has been used has a "playground" for the Semantic Web since 2007. The data is not centrally controlled or curated, which means you may notice that it has **inconsistent quality**. However, that is not of paramount importance to us - DBpedia data is still useful for demonstrating the concepts and functions behind SPARQL queries.

# YASGUI interface 

It is possible to query DBpedia directly from their endpoint at the URL above. However, we will instead use another SPARQL query editor called [**YASGUI**](http://yasgui.triply.cc) which has a few additional features which makes it easier to use. 

In addition, it is independent of the specific triplestore you want to query - whereas at the DBpedia SPARQL endpoint you can only query information from DBpedia itself. In contrast, you can **query any SPARQL endpoint with YASGUI** as long as you know its URL!

The link to the YASGUI interface is located here:

> **https://yasgui.triply.cc**

# YASGUI interface 

> **https://yasgui.triply.cc**

<img src="yasgui-interface.png">

In [2]:
# Set the SPARQL Kernel parameters
%endpoint http://dbpedia.org/sparql

# This is optional, it would increase the log level
%log debug

# Uncomment the next line to return label in english and avoid duplicates
# %lang en

Endpoint set to: http://dbpedia.org/sparql
Logging set to DEBUG

# SPARQL query components

As we saw in Lecture 4, variables in SPARQL are defined using a `?` symbol preceding it (e.g. `?my_variable`)

```sparql
# prefix declarations: for abbreviating IRIs
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
# graph IRI (optional): which RDF graph(s) are being queried
FROM
# query type: select, construct, insert, ask, describe etc. and which variables to return
SELECT *
# query pattern: specifying what to query for in the underlying dataset
WHERE {
    ?s ?p ?o .
}
# query modifiers: features which help you configure the presentation of the query results e.g. slicing, ordering, and rearranging
ORDER BY ?s
LIMIT 10
```

# 1. Run your first SPARQL query

Ask for **[all triples in DBpedia](http://yasgui.triply.cc/#query=SELECT%20*%0AWHERE%20%7B%0A%20%20%20%20%3Fsubject%20%3Fpredicate%20%3Fobject%20.%0A%7D&endpoint=https%3A%2F%2Fdbpedia.org%2Fsparql&requestMethod=POST&tabTitle=Query&headers=%7B%7D&contentTypeConstruct=text%2Fturtle%2C*%2F*%3Bq%3D0.9&contentTypeSelect=application%2Fsparql-results%2Bjson%2C*%2F*%3Bq%3D0.9&outputFormat=table)**:

In [2]:
SELECT *
WHERE {
    ?subject ?predicate ?object .
}

subject 
 predicate 
 object http://www.openlinksw.com/virtrdf-data-formats#default-iid 
 http://www.w3.org/1999/02/22-rdf-syntax-ns#type 
 http://www.openlinksw.com/schemas/virtrdf#QuadMapFormat http://www.openlinksw.com/virtrdf-data-formats#default-iid-nullable 
 http://www.w3.org/1999/02/22-rdf-syntax-ns#type 
 http://www.openlinksw.com/schemas/virtrdf#QuadMapFormat http://www.openlinksw.com/virtrdf-data-formats#default-iid-nonblank 
 http://www.w3.org/1999/02/22-rdf-syntax-ns#type 
 http://www.openlinksw.com/schemas/virtrdf#QuadMapFormat http://www.openlinksw.com/virtrdf-data-formats#default-iid-nonblank-nullable 
 http://www.w3.org/1999/02/22-rdf-syntax-ns#type 
 http://www.openlinksw.com/schemas/virtrdf#QuadMapFormat http://www.openlinksw.com/virtrdf-data-formats#default 
 http://www.w3.org/1999/02/22-rdf-syntax-ns#type 
 http://www.openlinksw.com/schemas/virtrdf#QuadMapFormat http://www.openlinksw.com/virtrdf-data-formats#default-nullable 
 http://www.w3.org/1999/02/22-rdf-syntax-ns#type 
 http://www.openlinksw.com/schemas/virtrdf#QuadMapFormat http://www.openlinksw.com/virtrdf-data-formats#sql-varchar 
 http://www.w3.org/1999/02/22-rdf-syntax-ns#type 
 http://www.openlinksw.com/schemas/virtrdf#QuadMapFormat http://www.openlinksw.com/virtrdf-data-formats#sql-varchar-nullable 
 http://www.w3.org/1999/02/22-rdf-syntax-ns#type 
 http://www.openlinksw.com/schemas/virtrdf#QuadMapFormat http://www.openlinksw.com/virtrdf-data-formats#sql-varchar-dt 
 http://www.w3.org/1999/02/22-rdf-syntax-ns#type 
 http://www.openlinksw.com/schemas/virtrdf#QuadMapFormat http://www.openlinksw.com/virtrdf-data-formats#sql-varchar-dt-nullable 
 http://www.w3.org/1999/02/22-rdf-syntax-ns#type 
 http://www.openlinksw.com/schemas/virtrdf#QuadMapFormat http://www.openlinksw.com/virtrdf-data-formats#sql-varchar-lang 
 http://www.w3.org/1999/02/22-rdf-syntax-ns#type 
 http://www.openlinksw.com/schemas/virtrdf#QuadMapFormat http://www.openlinksw.com/virtrdf-data-formats#sql-varchar-lang-nullable 
 http://www.w3.org/1999/02/22-rdf-syntax-ns#type 
 http://www.openlinksw.com/schemas/virtrdf#QuadMapFormat http://www.openlinksw.com/virtrdf-data-formats#sql-varchar-fixedlang-x-any 
 http://www.w3.org/1999/02/22-rdf-syntax-ns#type 
 http://www.openlinksw.com/schemas/virtrdf#QuadMapFormat http://www.openlinksw.com/virtrdf-data-formats#sql-varchar-fixedlang-x-any-nullable 
 http://www.w3.org/1999/02/22-rdf-syntax-ns#type 
 http://www.openlinksw.com/schemas/virtrdf#QuadMapFormat http://www.openlinksw.com/virtrdf-data-formats#sql-varchar-uri 
 http://www.w3.org/1999/02/22-rdf-syntax-ns#type 
 http://www.openlinksw.com/schemas/virtrdf#QuadMapFormat http://www.openlinksw.com/virtrdf-data-formats#sql-varchar-uri-nullable 
 http://www.w3.org/1999/02/22-rdf-syntax-ns#type 
 http://www.openlinksw.com/schemas/virtrdf#QuadMapFormat http://www.openlinksw.com/virtrdf-data-formats#sql-varchar-literal 
 http://www.w3.org/1999/02/22-rdf-syntax-ns#type 
 http://www.openlinksw.com/schemas/virtrdf#QuadMapFormat http://www.openlinksw.com/virtrdf-data-formats#sql-varchar-literal-nullable 
 http://www.w3.org/1999/02/22-rdf-syntax-ns#type 
 http://www.openlinksw.com/schemas/virtrdf#QuadMapFormat http://www.openlinksw.com/virtrdf-data-formats#sql-longvarchar 
 http://www.w3.org/1999/02/22-rdf-syntax-ns#type 
 http://www.openlinksw.com/schemas/virtrdf#QuadMapFormat http://www.openlinksw.com/virtrdf-data-formats#sql-longvarchar-nullable 
 http://www.w3.org/1999/02/22-rdf-syntax-ns#type 
 http://www.openlinksw.com/schemas/virtrdf#QuadMapFormat Total: 10000, Shown: 20

Observe that **DBpedia limits to 10.000 results by default** (and the SPARQL kernel shows only 20 for readability reasons). **Note:** DBpedia contains much more than 10.000 triples.

...but you may say: I don't know what these triples are about. They don't really describe anything I am interested in. You will see that many of these are triples that describe the endpoint itself and the technologies it uses. Let's filter these for more interesting information...

# 2. Get the instances of a class

Use the standard predicate `rdf:type` to get **[all the books in DBpedia 📚](https://yasgui.triply.cc/#query=PREFIX%20rdf%3A%20%3Chttp%3A%2F%2Fwww.w3.org%2F1999%2F02%2F22-rdf-syntax-ns%23%3E%0APREFIX%20dbo%3A%20%3Chttp%3A%2F%2Fdbpedia.org%2Fontology%2F%3E%0ASELECT%20*%0AWHERE%20%7B%0A%20%20%3Fbook%20rdf%3Atype%20dbo%3ABook%20.%0A%7D%0A&endpoint=https%3A%2F%2Fdbpedia.org%2Fsparql&requestMethod=POST&tabTitle=Query&headers=%7B%7D&contentTypeConstruct=text%2Fturtle%2C*%2F*%3Bq%3D0.9&contentTypeSelect=application%2Fsparql-results%2Bjson%2C*%2F*%3Bq%3D0.9&outputFormat=table)**

Copy and paste the query below into **[yasgui.triply.cc](https://yasgui.triply.cc)** (make sure the SPARQL endpoint is specified as http://dbpedia.org/sparql)

In [3]:
PREFIX dbo: <http://dbpedia.org/ontology/>
SELECT *
WHERE {
    ?book rdf:type dbo:Book .
}

book http://dbpedia.org/resource/Havana_Storm http://dbpedia.org/resource/The_Awful_German_Language http://dbpedia.org/resource/Urmonotheismus http://dbpedia.org/resource/1066_and_All_That http://dbpedia.org/resource/2010:_Odyssey_Two http://dbpedia.org/resource/401(k) http://dbpedia.org/resource/A_Crown_of_Swords http://dbpedia.org/resource/A_Dictionary_of_the_English_Language http://dbpedia.org/resource/Adultery http://dbpedia.org/resource/Alice's_Adventures_in_Wonderland http://dbpedia.org/resource/Anne_of_Green_Gables http://dbpedia.org/resource/Ante-Nicene_Fathers http://dbpedia.org/resource/Around_the_World_in_Eighty_Days http://dbpedia.org/resource/Astronomy http://dbpedia.org/resource/BBC_World_News http://dbpedia.org/resource/Bacteriology http://dbpedia.org/resource/Balthasar_Behem_Codex http://dbpedia.org/resource/Between_Planets http://dbpedia.org/resource/Beyond_This_Horizon http://dbpedia.org/resource/Biosalinity Total: 10000, Shown: 20

Here a prefix is defined for the DBpedia vocabulary / ontology:

```sparql
PREFIX dbo: <http://dbpedia.org/ontology/>
```

The `rdf:` prefix is defined by default, but `rdf:type` can be shortened to `a`

```sparql
SELECT * WHERE {
  ?book a dbo:Book .
}
```

If we execute this query on the following two triples only, the result would contain just `<http://book1>` (the only valid binding for ```?book ```)

```turtle
<http://book1> rdf:type <http://dbpedia.org/ontology/Book> .
<http://country1> rdf:type <http://dbpedia.org/ontology/Country> .
```

Experiment with variations of this query. I.e. modify it to return specific variables rather than ``*``. Do you notice a difference?

# 3. Get property of class

Get [**the author of each returned book** 🖋️](http://yasgui.triply.cc/#query=PREFIX%20dbo%3A%20%3Chttp%3A%2F%2Fdbpedia.org%2Fontology%2F%3E%0ASELECT%20*%0AWHERE%20%7B%0A%20%20%20%20%3Fbook%20a%20dbo%3ABook%20.%0A%20%20%20%20%3Fbook%20dbo%3Aauthor%20%3Fauthor%20.%0A%7D&endpoint=https%3A%2F%2Fdbpedia.org%2Fsparql&requestMethod=POST&tabTitle=Query&headers=%7B%7D&contentTypeConstruct=text%2Fturtle%2C*%2F*%3Bq%3D0.9&contentTypeSelect=application%2Fsparql-results%2Bjson%2C*%2F*%3Bq%3D0.9&outputFormat=table) (when an author is defined):

In [5]:
PREFIX dbo: <http://dbpedia.org/ontology/>
SELECT *
WHERE {
    ?book a dbo:Book .
    ?book dbo:author ?author .
}

book 
 author http://dbpedia.org/resource/1066_and_All_That 
 http://dbpedia.org/resource/R._J._Yeatman http://dbpedia.org/resource/1066_and_All_That 
 http://dbpedia.org/resource/W._C._Sellar http://dbpedia.org/resource/2010:_Odyssey_Two 
 http://dbpedia.org/resource/Arthur_C._Clarke http://dbpedia.org/resource/A_Crown_of_Swords 
 http://dbpedia.org/resource/Robert_Jordan http://dbpedia.org/resource/A_Dictionary_of_the_English_Language 
 http://dbpedia.org/resource/Samuel_Johnson http://dbpedia.org/resource/Alice's_Adventures_in_Wonderland 
 http://dbpedia.org/resource/Lewis_Carroll http://dbpedia.org/resource/Anne_of_Green_Gables 
 http://dbpedia.org/resource/Lucy_Maud_Montgomery http://dbpedia.org/resource/Around_the_World_in_Eighty_Days 
 http://dbpedia.org/resource/Jules_Verne http://dbpedia.org/resource/Between_Planets 
 http://dbpedia.org/resource/Robert_A._Heinlein http://dbpedia.org/resource/Beyond_This_Horizon 
 http://dbpedia.org/resource/Robert_A._Heinlein http://dbpedia.org/resource/Blood_Music_(novel) 
 http://dbpedia.org/resource/Greg_Bear http://dbpedia.org/resource/Brave_New_World 
 http://dbpedia.org/resource/Aldous_Huxley http://dbpedia.org/resource/Castle_Rackrent 
 http://dbpedia.org/resource/Maria_Edgeworth http://dbpedia.org/resource/Catch-22 
 http://dbpedia.org/resource/Joseph_Heller http://dbpedia.org/resource/Charlie_and_the_Chocolate_Factory 
 http://dbpedia.org/resource/Roald_Dahl http://dbpedia.org/resource/Citizen_of_the_Galaxy 
 http://dbpedia.org/resource/Robert_A._Heinlein http://dbpedia.org/resource/Clash_of_Civilizations 
 http://dbpedia.org/resource/Samuel_P._Huntington http://dbpedia.org/resource/Clear_and_Present_Danger 
 http://dbpedia.org/resource/Tom_Clancy http://dbpedia.org/resource/Commentarii_de_Bello_Gallico 
 http://dbpedia.org/resource/Julius_Caesar http://dbpedia.org/resource/Commentarii_de_Bello_Gallico 
 http://dbpedia.org/resource/Aulus_Hirtius Total: 10000, Shown: 20

A Turtle-like syntax can also be used in SPARQL to make the query more readable:

```sparql
PREFIX dbo: <http://dbpedia.org/ontology/>
SELECT *
WHERE {
    ?book a dbo:Book ; 
        dbo:author ?author .
}
```

Consider a graph with the following 4 statements:

```turtle
<http://book1> rdf:type <http://dbpedia.org/ontology/Book> .
<http://book1> dbo:author <http://author1> .
<http://book2> rdf:type <http://dbpedia.org/ontology/Book> .
<http://book2> dbo:contributor <http://author2> .
```

The previous query will return only **one row of results with `<http://book1>` and `<http://author1>`**

# Presentation and processing of SPARQL query results

* `LIMIT`: specify the maximal number of results (table rows) 
* `OFFSET`: position of the first delivered result SELECT 
* `DISTINCT`: removal of duplicate table rows

> `LIMIT` and `OFFSET` are often used in conjunction with `ORDER BY`


# 4. Retrieve specific variables and limit

You can specify the variables you want to be returned. And `LIMIT` the number of results returned.

This query will return 10 authors of books in DBpedia, with duplicates if the author wrote multiple books 👥

In [5]:
PREFIX dbo: <http://dbpedia.org/ontology/>
SELECT ?author
WHERE {
    ?book a dbo:Book ;
        dbo:author ?author .
} 
LIMIT 10

author http://dbpedia.org/resource/R._J._Yeatman http://dbpedia.org/resource/W._C._Sellar http://dbpedia.org/resource/Arthur_C._Clarke http://dbpedia.org/resource/Robert_Jordan http://dbpedia.org/resource/Samuel_Johnson http://dbpedia.org/resource/Lewis_Carroll http://dbpedia.org/resource/Lucy_Maud_Montgomery http://dbpedia.org/resource/Jules_Verne http://dbpedia.org/resource/Robert_A._Heinlein http://dbpedia.org/resource/Robert_A._Heinlein Total: 10, Shown: 10

# 5. Retrieve distinct results

You can also enforce that only unique `DISTINCT` results are returned

> Use the `*` if you want all variables

This query will return all authors of books in DBpedia **without duplicates**, even if they wrote multiple books 👤

In [6]:
PREFIX dbo: <http://dbpedia.org/ontology/>
SELECT DISTINCT ?author
WHERE {
    ?book a dbo:Book ;
        dbo:author ?author .
} LIMIT 10

author http://dbpedia.org/resource/R._J._Yeatman http://dbpedia.org/resource/W._C._Sellar http://dbpedia.org/resource/Arthur_C._Clarke http://dbpedia.org/resource/Robert_Jordan http://dbpedia.org/resource/Samuel_Johnson http://dbpedia.org/resource/Lewis_Carroll http://dbpedia.org/resource/Lucy_Maud_Montgomery http://dbpedia.org/resource/Jules_Verne http://dbpedia.org/resource/Robert_A._Heinlein http://dbpedia.org/resource/Greg_Bear Total: 10, Shown: 10

# Your turn 🎬

Use **[https://yasgui.triply.cc](https://yasgui.triply.cc)** to write and run SPARQL queries on the DBpedia RDF graph to retrieve the following answers below:

1. Get the list of movies (`dbo:Film`) without duplicates 🎥

2. Get the list of movies (`dbo:Film`) and the actors who star in this movie (`dbo:starring`)

3. Get the label of the movie (return `?movieLabel ?actor`, not the movie URI)

1. Get the list of movies (`dbo:Film`) without duplicates

In [25]:
PREFIX dbo: <http://dbpedia.org/ontology/>
SELECT DISTINCT ?movie
WHERE {
    ?movie a dbo:Film .
} 

movie http://dbpedia.org/resource/4000_Miles http://dbpedia.org/resource/The_Man_Who_Came_Back http://dbpedia.org/resource/12_Monkeys http://dbpedia.org/resource/Airplane! http://dbpedia.org/resource/American_Beauty_(1999_film) http://dbpedia.org/resource/Amélie http://dbpedia.org/resource/Andrzej_Wajda http://dbpedia.org/resource/Animal_Crackers_(1930_film) http://dbpedia.org/resource/Animated_cartoon http://dbpedia.org/resource/Anne_of_the_Thousand_Days http://dbpedia.org/resource/Anywhere_but_Here_(film) http://dbpedia.org/resource/Armageddon_(1998_film) http://dbpedia.org/resource/Arms_and_the_Man http://dbpedia.org/resource/Balducci_levitation http://dbpedia.org/resource/Beautiful_Girls_(film) http://dbpedia.org/resource/Berry http://dbpedia.org/resource/Black_Narcissus http://dbpedia.org/resource/Black_tie http://dbpedia.org/resource/Breakfast_at_Tiffany's_(film) http://dbpedia.org/resource/Carry_On_(franchise) Total: 10000, Shown: 20

2. Get the list of movies (`dbo:Film`) and the actors who star in this movie (`dbo:starring`)

In [23]:
PREFIX dbo: <http://dbpedia.org/ontology/>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
SELECT DISTINCT ?movie ?actor
WHERE {
    ?movie a dbo:Film ;
        dbo:starring ?actor .
} 
LIMIT 10

movie 
 actor http://dbpedia.org/resource/12_Monkeys 
 http://dbpedia.org/resource/Brad_Pitt http://dbpedia.org/resource/12_Monkeys 
 http://dbpedia.org/resource/Madeleine_Stowe http://dbpedia.org/resource/12_Monkeys 
 http://dbpedia.org/resource/Bruce_Willis http://dbpedia.org/resource/12_Monkeys 
 http://dbpedia.org/resource/Christopher_Plummer http://dbpedia.org/resource/Amélie 
 http://dbpedia.org/resource/Audrey_Tautou http://dbpedia.org/resource/Amélie 
 http://dbpedia.org/resource/Mathieu_Kassovitz http://dbpedia.org/resource/Animal_Crackers_(1930_film) 
 http://dbpedia.org/resource/Zeppo_Marx http://dbpedia.org/resource/Animal_Crackers_(1930_film) 
 http://dbpedia.org/resource/Groucho_Marx http://dbpedia.org/resource/Animal_Crackers_(1930_film) 
 http://dbpedia.org/resource/Chico_Marx http://dbpedia.org/resource/Animal_Crackers_(1930_film) 
 http://dbpedia.org/resource/Harpo_Marx Total: 10, Shown: 10

3. Get the label of the movie (return `?movieLabel ?actor`, not the movie URI)

In [24]:
PREFIX dbo: <http://dbpedia.org/ontology/>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
SELECT DISTINCT ?movieLabel ?actor
WHERE {
    ?movie a dbo:Film ;
        dbo:starring ?actor ;
        rdfs:label ?movieLabel .
} 
LIMIT 10

movieLabel 
 actor 12 Monkeys 
 http://dbpedia.org/resource/Brad_Pitt 12 قردا 
 http://dbpedia.org/resource/Brad_Pitt 12 Monkeys 
 http://dbpedia.org/resource/Brad_Pitt Doce monos 
 http://dbpedia.org/resource/Brad_Pitt L'Armée des douze singes 
 http://dbpedia.org/resource/Brad_Pitt L'esercito delle 12 scimmie 
 http://dbpedia.org/resource/Brad_Pitt 12モンキーズ 
 http://dbpedia.org/resource/Brad_Pitt 12 Monkeys (film) 
 http://dbpedia.org/resource/Brad_Pitt 12 małp 
 http://dbpedia.org/resource/Brad_Pitt Twelve Monkeys 
 http://dbpedia.org/resource/Brad_Pitt Total: 10, Shown: 10

# FILTER the results

## Comparison operators: <, =, >, <=, >=, !=
* Comparison of data literals according to natural order
* Support for numerical data types, xsd:dateTime, xsd:string (alphabetic ordering), xsd:Boolean (1>0)
* For other types and other RDF-elements, only = and != are available
* Comparison of literals of incompatible types (e.g. xsd:string and xsd:integer) is not allowed, they must be converted

## Arithmetic operators: +, -, *, /
* Support for numerical data types
* Used to combine values in filter conditions 
* E.g. `FILTER(?weight/ (?size*?size)>=25)`


# 6. Filter by number of pages

This query will return 10 authors who wrote a book with more than 500 pages 📖

In [7]:
PREFIX dbo: <http://dbpedia.org/ontology/>
SELECT DISTINCT ?author
WHERE {
    ?book a dbo:Book ;
        dbo:author ?author ;
        dbo:numberOfPages ?pages .
    FILTER (?pages > 500)
} LIMIT 10

author http://dbpedia.org/resource/Robert_Jordan http://dbpedia.org/resource/Tom_Clancy http://dbpedia.org/resource/Douglas_Hofstadter http://dbpedia.org/resource/Richard_Llewellyn http://dbpedia.org/resource/Charles_Darwin http://dbpedia.org/resource/Henry_Martyn_Robert http://dbpedia.org/resource/Iain_Banks http://dbpedia.org/resource/Victor_Hugo http://dbpedia.org/resource/Robert_Anton_Wilson http://dbpedia.org/resource/Robert_Shea Total: 10, Shown: 10

# 7. Filter books on name content

* Cast the variable `?name` using the `str()` function, to avoid datatypes incompatibility
* Use the **function** `contains(string_to_look_in,string_to_look_for)` to only keep books with the string "grand" in their label 🏷️

In [10]:
PREFIX dbo:<http://dbpedia.org/ontology/>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
SELECT ?name ?author 
WHERE {
    ?book a dbo:Book ;
        dbo:author ?author ;
        rdfs:label ?name . 
    FILTER (contains(str(?name),"grand"))
} LIMIT 10

name 
 author Il grande divorzio 
 http://dbpedia.org/resource/C._S._Lewis La grande caccia 
 http://dbpedia.org/resource/Robert_Jordan La grande fuga dell'Ottobre Rosso 
 http://dbpedia.org/resource/Tom_Clancy La grande storia del tempo 
 http://dbpedia.org/resource/Leonard_Mlodinow La grande storia del tempo 
 http://dbpedia.org/resource/Stephen_Hawking Il grande ritratto 
 http://dbpedia.org/resource/Dino_Buzzati Il piccolo grande uomo (romanzo) 
 http://dbpedia.org/resource/Thomas_Berger_(novelist) Il grande mare dei Sargassi 
 http://dbpedia.org/resource/Jean_Rhys La Conspiration au grand jour 
 http://dbpedia.org/resource/H._G._Wells Considérations sur les causes de la grandeur des Romains et de leur décadence 
 http://dbpedia.org/resource/Montesquieu Total: 10, Shown: 10

# Using conditions in FILTER

Filter conditions can be linked with boolean operators: `&&`, `||`, `!`

> Keep when pages superior to 500 and does not contains "grand" in its name

```sparql
FILTER((?pages > 500) && !contains(?name,"grand"))
```

> Keep when pages superior to 500 or contains "grand" in its name

```sparql
FILTER((?pages > 500) || contains(?name,"grand"))
```

Expressible through graph pattern:
* **Conjunction** corresponds to specifications of **several filters**.
* **Disjunction** corresponds to application of filters in **alternative patterns** (use `UNION`).


# 8. Get the language of a literal

We can use the `lang()` function in the `SELECT` statement to return the language of the book label 💬

> Give it a name using the keyword `as`

In [13]:
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
PREFIX dbo:<http://dbpedia.org/ontology/>
SELECT ?name ?author lang(?name) as ?lang
WHERE  
    ?book a dbo:Book ;
        dbo:author ?author ;
        rdfs:label ?name . 
} LIMIT 10

name 
 author 
 lang 1066 and All That 
 http://dbpedia.org/resource/R._J._Yeatman 
 en 1066 and All That 
 http://dbpedia.org/resource/R._J._Yeatman 
 es 1066 and All That 
 http://dbpedia.org/resource/R._J._Yeatman 
 fr 1066 and All That 
 http://dbpedia.org/resource/W._C._Sellar 
 en 1066 and All That 
 http://dbpedia.org/resource/W._C._Sellar 
 es 1066 and All That 
 http://dbpedia.org/resource/W._C._Sellar 
 fr 2010: Odyssey Two 
 http://dbpedia.org/resource/Arthur_C._Clarke 
 en Odyssee 2010 – Das Jahr, in dem wir Kontakt aufnehmen 
 http://dbpedia.org/resource/Arthur_C._Clarke 
 de 2010: Odisea dos 
 http://dbpedia.org/resource/Arthur_C._Clarke 
 es 2010 : Odyssée deux 
 http://dbpedia.org/resource/Arthur_C._Clarke 
 fr Total: 10, Shown: 10

# Return only Dutch book titles

Use the `langMatches()` function to filter books with a label value Dutch 🇳🇱

In [14]:
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
PREFIX dbo:<http://dbpedia.org/ontology/>
SELECT ?name ?author
WHERE {
    ?book a dbo:Book ;
        dbo:author ?author ;
        rdfs:label ?name . 
    FILTER (langMatches(lang(?name),"nl"))
} LIMIT 10

name 
 author 2010: Odyssey Two 
 http://dbpedia.org/resource/Arthur_C._Clarke Een Kroon van Zwaarden 
 http://dbpedia.org/resource/Robert_Jordan A Dictionary of the English Language 
 http://dbpedia.org/resource/Samuel_Johnson Alice's Adventures in Wonderland 
 http://dbpedia.org/resource/Lewis_Carroll Anne van het Groene Huis 
 http://dbpedia.org/resource/Lucy_Maud_Montgomery De reis om de wereld in tachtig dagen 
 http://dbpedia.org/resource/Jules_Verne Tussen de planeten 
 http://dbpedia.org/resource/Robert_A._Heinlein Brave New World (roman) 
 http://dbpedia.org/resource/Aldous_Huxley Catch-22 (roman) 
 http://dbpedia.org/resource/Joseph_Heller Sjakie en de chocoladefabriek 
 http://dbpedia.org/resource/Roald_Dahl Total: 10, Shown: 10

# 9. Sorting results

We can use **[`ORDER BY`](https://yasgui.triply.cc/#query=PREFIX%20dbo%3A%3Chttp%3A%2F%2Fdbpedia.org%2Fontology%2F%3E%0APREFIX%20rdfs%3A%20%3Chttp%3A%2F%2Fwww.w3.org%2F2000%2F01%2Frdf-schema%23%3E%0ASELECT%20%3Fname%20%3Fauthor%20%3Fpages%20%3Fbook%0AWHERE%20%7B%0A%20%20%20%20%3Fbook%20a%20dbo%3ABook%20%3B%0A%20%20%20%20%20%20%20%20dbo%3Aauthor%20%3Fauthor%20%3B%0A%20%20%20%20%20%20%20%20rdfs%3Alabel%20%3Fname%20%3B%0A%20%20%20%20%20%20%20%20dbo%3AnumberOfPages%20%3Fpages%20.%0A%7D%20%0AORDER%20BY%20%3Fpages%0ALIMIT%2010&endpoint=https%3A%2F%2Fdbpedia.org%2Fsparql&requestMethod=POST&tabTitle=Query&headers=%7B%7D&contentTypeConstruct=text%2Fturtle%2C*%2F*%3Bq%3D0.9&contentTypeSelect=application%2Fsparql-results%2Bjson%2C*%2F*%3Bq%3D0.9&outputFormat=table)**, to order the books by the number of pages it contains in ascending order by default 📤

In [15]:
PREFIX dbo:<http://dbpedia.org/ontology/>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
SELECT ?name ?author ?pages
WHERE {
    ?book a dbo:Book ;
        dbo:author ?author ;
        rdfs:label ?name ;
        dbo:numberOfPages ?pages .
} 
ORDER BY ?pages
LIMIT 10

name 
 author 
 pages Liberia; or, Mr. Peyton's Experiments 
 http://dbpedia.org/resource/Sarah_Josepha_Hale 
 0 The Cabin and Parlor; or, Slaves and Masters 
 http://dbpedia.org/resource/Charles_Jacobs_Peterson 
 0 The Lofty and the Lowly, or Good in All and None All Good 
 http://dbpedia.org/resource/Maria_Jane_McIntosh 
 0 Geological Evidences of the Antiquity of Man 
 http://dbpedia.org/resource/Charles_Lyell 
 0 Geological Evidences of the Antiquity of Man 
 http://dbpedia.org/resource/Charles_Lyell 
 0 The Vampyre 
 http://dbpedia.org/resource/John_William_Polidori 
 0 Der Vampyr (Polidori) 
 http://dbpedia.org/resource/John_William_Polidori 
 0 El vampiro (1819) 
 http://dbpedia.org/resource/John_William_Polidori 
 0 Le Vampire (nouvelle) 
 http://dbpedia.org/resource/John_William_Polidori 
 0 Il vampiro (racconto) 
 http://dbpedia.org/resource/John_William_Polidori 
 0 Total: 10, Shown: 10

# Order by results descending 

Order results from books with the highest number of pages to the lowest 📥

In [16]:
PREFIX dbo:<http://dbpedia.org/ontology/>
PREFIX dbp:<http://dbpedia.org/property/>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
SELECT ?name ?author ?pages
WHERE {
    ?book a dbo:Book ;
        dbo:author ?author ;
        rdfs:label ?name ;
        dbo:numberOfPages ?pages .
    FILTER (langMatches(lang(?name),"en"))
} 
ORDER BY DESC(?pages)
LIMIT 10

name 
 author 
 pages The Tolkien Reader 
 http://dbpedia.org/resource/J._R._R._Tolkien 
 241127964 L. Ron Hubbard, Messiah or Madman? 
 http://dbpedia.org/resource/Ronald_DeWolf 
 1992460 L. Ron Hubbard, Messiah or Madman? 
 http://dbpedia.org/resource/Bent_Corydon 
 1992460 L. Ron Hubbard, Messiah or Madman? 
 http://dbpedia.org/resource/Ronald_DeWolf 
 1987402 L. Ron Hubbard, Messiah or Madman? 
 http://dbpedia.org/resource/Bent_Corydon 
 1987402 The Merlin Conspiracy 
 http://dbpedia.org/resource/Diana_Wynne_Jones 
 473468 Political Liberalism 
 http://dbpedia.org/resource/John_Rawls 
 435576 Halo: The Flood 
 http://dbpedia.org/resource/William_C._Dietz 
 352432 Halo: The Fall of Reach 
 http://dbpedia.org/resource/Eric_Nylund 
 352416 Unsung: A History of Women in American Music 
 http://dbpedia.org/resource/Christine_Ammer 
 317382 Total: 10, Shown: 10

# 10. Count results

Count the number of books

In [26]:
PREFIX dbo:<http://dbpedia.org/ontology/>
SELECT (count(?book) as ?book_count)
WHERE {
    ?book a dbo:Book .
}

book_count 64239 Total: 1, Shown: 1

# Count a specific pattern

We can also count how many row of a pattern are returned

Count the number of books for each author 📚

In [27]:
PREFIX dbo:<http://dbpedia.org/ontology/>
SELECT ?author (count(?book) as ?book_count)
WHERE {
    ?book a dbo:Book ;
        dbo:author ?author .
} LIMIT 10

author 
 book_count http://dbpedia.org/resource/Daniel_Carter_Beard 
 1 http://dbpedia.org/resource/Raymond_Smullyan 
 1 http://dbpedia.org/resource/William_Donaldson 
 1 http://dbpedia.org/resource/Joe_Conason 
 1 http://dbpedia.org/resource/John_Lennon 
 3 http://dbpedia.org/resource/Jürgen_Habermas 
 8 http://dbpedia.org/resource/Robert_Nathan 
 1 http://dbpedia.org/resource/J._P._Martin 
 6 http://dbpedia.org/resource/Anthony_Everitt 
 1 http://dbpedia.org/resource/Arthur_R.G._Solmssen 
 1 Total: 10, Shown: 10

# 11. Return results from different patterns

Use `UNION` to return the books name wherever it is defined with `rdfs:label` or `dbp:name`:

In [31]:
PREFIX dbp: <http://dbpedia.org/property/>
PREFIX dbo:<http://dbpedia.org/ontology/>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
SELECT ?book ?name
WHERE {
    {
        ?book a dbo:Book ;
            rdfs:label ?name .
        FILTER (langMatches(lang(?name),"en"))
    } UNION {
        ?book a dbo:Book ;
            dbp:name ?name .
    }
    FILTER (contains(str(?name), "grand"))
}

book 
 name http://dbpedia.org/resource/Alphabetum_grandonico-malabaricum_sive_samscrudonicum 
 Alphabetum grandonico-malabaricum sive samscrudonicum http://dbpedia.org/resource/Les_grands_cimetières_sous_la_lune 
 Les grands cimetières sous la lune http://dbpedia.org/resource/L'Art_d'être_grand-père 
 L'Art d'être grand-père http://dbpedia.org/resource/La_Cour_des_grands 
 La Cour des grands http://dbpedia.org/resource/Plame_affair_grand_jury_investigation 
 Plame affair grand jury investigation http://dbpedia.org/resource/Servitude_et_grandeur_militaires 
 Servitude et grandeur militaires http://dbpedia.org/resource/My_Grandfather_Would_Have_Shot_Me 
 Jennifer Teege , My grandfather would have shot me : a Black woman discovers her family's Nazi past Total: 7, Shown: 7

# Without UNION

For `rdfs:label` only 6 results

> Check how many results for `dbp:name`

In [32]:
PREFIX dbp: <http://dbpedia.org/property/>
PREFIX dbo:<http://dbpedia.org/ontology/>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
SELECT ?book ?name
WHERE {
    ?book a dbo:Book ;
        rdfs:label ?name .
    FILTER (langMatches(lang(?name),"en"))
    FILTER (contains(str(?name), "grand"))
}

book 
 name http://dbpedia.org/resource/Alphabetum_grandonico-malabaricum_sive_samscrudonicum 
 Alphabetum grandonico-malabaricum sive samscrudonicum http://dbpedia.org/resource/Les_grands_cimetières_sous_la_lune 
 Les grands cimetières sous la lune http://dbpedia.org/resource/L'Art_d'être_grand-père 
 L'Art d'être grand-père http://dbpedia.org/resource/La_Cour_des_grands 
 La Cour des grands http://dbpedia.org/resource/Plame_affair_grand_jury_investigation 
 Plame affair grand jury investigation http://dbpedia.org/resource/Servitude_et_grandeur_militaires 
 Servitude et grandeur militaires Total: 6, Shown: 6

# Search on DBpedia 🔎

Use **[https://yasgui.triply.cc](https://yasgui.triply.cc)** to write and run SPARQL queries on the DBpedia RDF graph to retrieve the following answers below:

1. List 10 authors and the capital cities ([dbo:capital](http://dbpedia.org/ontology/capital)) of the countries in which they were born ([dbo:birthPlace](http://dbpedia.org/ontology/birthPlace))

2. The number of authors for the book that has the **English** title: "1066 and All That"

3. All books with a name in **English** starting with "http" (ignore the case)

4. A list of all the unique book categories (dct:subject) for all short books (less than 300 pages) written by authors who were born in Amsterdam

* **Search for functions in the specifications: https://www.w3.org/TR/sparql11-query**
* How do I find vocabulary to use in my SPARQL query from DBpedia? Search in **[DBpedia ontology classes](http://mappings.dbpedia.org/server/ontology/classes/)**, or on google, e.g., search for: "**[dbpedia capital](https://www.google.com/search?&q=dbpedia+capital)**"


# Examples of other public SPARQL endpoints 🔗

* Wikidata, facts powering Wikipedia infobox: https://query.wikidata.org/sparql
* Bio2RDF, linked data for the life sciences: https://bio2rdf.org/sparql
* Disgenet, gene-disease association: http://rdf.disgenet.org/sparql
* PathwayCommons, resource for biological pathways analysis: http://rdf.pathwaycommons.org/sparql
* EU publications office, court decisions and legislative documents from the EU: http://publications.europa.eu/webapi/rdf/sparql
* Finland legal open data, cases and legislation: https://data.finlex.fi/en/sparql 

# Other interesting links

* Wikidata SPARQL queries around the SARS-CoV-2 virus and pandemic: https://egonw.github.io/SARS-CoV-2-Queries
* Use [prefix.cc](http://prefix.cc/) to discover the full IRIs for unknown prefixes you may encounter